056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,131 @@
+
+### Rollability checks
+ - Attempts to connect an admin client to the broker and if it can't connect rolls the broker
+ - Does not force roll a broker performing log recovery
+ - Does not roll a broker if doing so would take the in-sync replicas count below `min.insync.replicas`
+
+### Configuration changes
+ - Retrieves the current Kafka configurations of the broker via admin client and compares it with the desired configurations specified in the Kafka CR. 
+ - Performs dynamic configuration updates if possible, otherwise rolls broker on configuration change
+
+In KRaft mode the KafkaRoller currently skips controller only nodes, but performs the above steps on any combined or broker only nodes.
+This is causing a problem in combined mode because if the quorum has not formed due to some of the nodes not being ready 
+the KafkaRoller will wait still try to contact the broker via the admin client.
+This call fails because the quorum is not formed, so in some cases this results in the cluster being stuck with some nodes 
+in a pending state.
+
+## Motivation
+
+When running in the KRaft mode the controller pods need to be rolled if configuration changes occur, also at the moment the 
+existing logic is blocking the full implementation of liveness and readiness checks as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md).
+
+## Proposal
+
+The KafkaRoller behaviour should be unchanged when operating against a ZooKeeper based cluster.
+
+The behaviour when operating against a KRaft cluster is described below.
+
+### Order to roll nodes
+Roll in order:
+- Unready controller/combined nodes
+- Ready controller/combined nodes
+- Active controller
+- Unready broker-only nodes
+- Ready broker-only nodes
+
+
+### Triggers
+The following are some of the triggers that would roll a KRaft controller or combined node:
+- Pod is annotated for manual rolling update
+- Pod is unready because it is in `CrashLoopBackOff`, `ImagePullBackOff`, `ContainerCreating` or `Pending` and `Unschedulable`, 
+- Controller's configuration has changed and cannot be updated dynamically.
+
+The triggers for broker remain the same as Zookeeper mode.
+
+**NOTE** There is a related proposal describing how to diff controller configuration currently being discussed: [PR #82](https://github.com/strimzi/proposals/pull/82)
+
+### Rollability checks
+
+The checks made by the KafkaRoller in different modes is described below.
+The checks include a new check for controllers to verify that rolling the node does not affect the quorum health.
+The proposed check is:
+- Fetch `lastCaughtUpTimestamp` for every controller node. `lastCaughtUpTimestamp` is the last millisecond timestamp at which a replica controller was known to be caught up with the quorum leader
+- Identify the `lastCaughtUpTimestamp` of the node that is the quorum leader .
+- Retrieve the current value of Kafka configuration `controller.quorum.fetch.timeout.ms`
```

Comment:
How / from where will it be retrieved?

Reply:
This is tricky to answer actually,  once [Proposal in PR #82](https://github.com/strimzi/proposals/pull/82) is implemented, we would be able to get the current value for
this configuration. Until then we would have to use the hard-coded default value which I'm not sure how acceptable is. I'm hoping both proposals would be implemented around similar time.

Reply:
As [Proposal in PR #82](https://github.com/strimzi/proposals/pull/82) is closed, we have changed the proposal to get the value of this property from the desired configurations instead. https://github.com/orgs/strimzi/discussions/9137 also discusses the justification for this in detail.